Lib/Cache/Engine/FileEngine.php

```diff
@@ -0,0 +1,462 @@
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         CakePHP(tm) v 1.2.0.4933
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+
+/**
+ * File Storage engine for cache. Filestorage is the slowest cache storage
+ * to read and write. However, it is good for servers that don't have other storage
+ * engine available, or have content which is not performance sensitive.
+ *
+ * You can configure a FileEngine cache, using Cache::config()
+ *
+ * @package       Cake.Cache.Engine
+ */
+class FileEngine extends CacheEngine {
+
+/**
+ * Instance of SplFileObject class
+ *
+ * @var File
+ */
+	protected $_File = null;
+
+/**
+ * Settings
+ *
+ * - path = absolute path to cache directory, default => CACHE
+ * - prefix = string prefix for filename, default => cake_
+ * - lock = enable file locking on write, default => true
+ * - serialize = serialize the data, default => true
+ *
+ * @var array
+ * @see CacheEngine::__defaults
+ */
+	public $settings = array();
+
+/**
+ * True unless FileEngine::__active(); fails
+ *
+ * @var bool
+ */
+	protected $_init = true;
+
+/**
+ * Initialize the Cache Engine
+ *
+ * Called automatically by the cache frontend
+ * To reinitialize the settings call Cache::engine('EngineName', [optional] settings = array());
+ *
+ * @param array $settings array of setting for the engine
+ * @return bool True if the engine has been successfully initialized, false if not
+ */
+	public function init($settings = array()) {
+		$settings += array(
+			'engine' => 'File',
+			'path' => CACHE,
+			'prefix' => 'cake_',
+			'lock' => true,
+			'serialize' => true,
+			'isWindows' => false,
+			'mask' => 0664
+		);
+		parent::init($settings);
+
+		if (DS === '\\') {
+			$this->settings['isWindows'] = true;
+		}
+		if (substr($this->settings['path'], -1) !== DS) {
+			$this->settings['path'] .= DS;
+		}
+		if (!empty($this->_groupPrefix)) {
+			$this->_groupPrefix = str_replace('_', DS, $this->_groupPrefix);
+		}
+		return $this->_active();
+	}
+
+/**
+ * Garbage collection. Permanently remove all expired and deleted data
+ *
+ * @param int $expires [optional] An expires timestamp, invalidating all data before.
+ * @return bool True if garbage collection was successful, false on failure
+ */
+	public function gc($expires = null) {
+		return $this->clear(true);
+	}
+
+/**
+ * Write data for key into cache
+ *
+ * @param string $key Identifier for the data
+ * @param mixed $data Data to be cached
+ * @param int $duration How long to cache the data, in seconds
+ * @param string|array $parentKey Unused.
+ * @return bool True if the data was successfully cached, false on failure
+ */
+	public function write($key, $data, $duration, $parentKey = '') {
```

Comment:
@jec3 So `$parentKey` is ignored here but you need to override the class and method to respect the new `write()` signature?

Reply:
Correct, because this implements the abstract class CacheEngine it must match the arguments specified there, which includes `$parentKey`.